### PR TITLE
adding mqtt client

### PR DESCRIPTION
**Pull request by marcusthebrown**
## Status
**READY**

## JIRA Ticket
SPACON-41

## Description
Add mqtt client and refactor network service.

## Todos
- [x] Tests
- [x] Documentation
- [x] License

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.

```sh
git fetch --all
git checkout <feature_branch> 
./gradlew connectedCheck
```

@boundlessgeo/spatial-connect

### CODE REVIEW

File: spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/mqtt/MqttHandler.java [hunk 1]
```diff
@@ -0,0 +1,214 @@
...
+        Log.d(LOG_TAG, "Lost connection to mqtt broker.", cause);
+        clientConnected.onNext(0);
+    }
+
+    @Override
+    public void messageArrived(String topic, MqttMessage message) throws Exception {
+        Log.v(LOG_TAG, "received message on topic " + topic);
+        SCMessageOuterClass.SCMessage scMessage = SCMessageOuterClass.SCMessage.parseFrom(message.getPayload());
+        Log.v(LOG_TAG, "message payload: " + scMessage.getPayload());
+        HashMap<String, SCMessageOuterClass.SCMessage> map = new HashMap<>();
```
Comment (marcusthebrown) [inline spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/mqtt/MqttHandler.java]: Use array of size 2 instead

File: spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/services/SCBackendService.java [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+ * SCBackendService handles any communication with backend SpatialConnect services.
+ */
+public class SCBackendService extends SCService {
+
+    private static final String LOG_TAG = SCBackendService.class.getSimpleName();
+    private static Context context;
+    private static MqttHandler mqttHandler;
+    private static HttpHandler httpHandler;
+
+    public static BehaviorSubject<Integer> initialized = BehaviorSubject.create(0);
```
Comment (marcusthebrown) [inline spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/services/SCBackendService.java]: rename to `running`

File: spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/services/SCConfigService.java [hunk 3]
```diff
@@ -245,10 +204,14 @@ public void addConfig(File configFile) {
     public void start() {
         Log.d(LOG_TAG, "Starting SCConfig Service.  Loading all configs");
         loadConfigs();
-        if (API_URL != null) {
-            loadRemoteConfig();
-            registerDevice();
-        }
+        SCBackendService.initialized.subscribe(new Action1<Integer>() {
```
Comment (marcusthebrown) [inline spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/services/SCConfigService.java]: This needs to work like the ios sdk, where the backend service is started, and after authenticating, the remote config is fetched from the mqtt topic

File: spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/services/SCBackendService.java [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+
+import okhttp3.Response;
+import rx.Observable;
+import rx.functions.Func1;
+import rx.subjects.BehaviorSubject;
+
+/**
+ * SCBackendService handles any communication with backend SpatialConnect services.
+ */
+public class SCBackendService extends SCService {
```
Comment (marcusthebrown) [inline spatialconnect/src/main/java/com/boundlessgeo/spatialconnect/services/SCBackendService.java]: override start method

**Comment by georific** [comment]
Reopening to merge

**Pull request merged**